Add support for more cloud events

afrittoli opened the pull request:
# Changes

This PR is split into two commits, to help with the review. 

Add support for more cloud events

Change the event type string for taskruns from dev.tekton.event.task.*
to dev.tekton.event.taskrun.*. Since cloud events are only sent by
pipeline resources - which are alpha - we don't need to comply with
the beta deprecation policy for this.

Extend the cloudevents module to include more event types for TaskRuns
and event types for PipelineRuns, with unit test coverage.
This is achieved by introducing the RunsToCompletion and
RunsToCompletionStatus interface which are met by TaskRun, PipelineRun
and their respective status types.

At this stage there is no event producer that generates these new
events. This is preparing the stage for the next changes where
we will start to emit cloud events in addition to the k8s events
we emit today.

No extra docs yet since the new events are not produced by anything yet.

Partially addresses #2082
Partially addresses #2684

The second:

Make phase condition reasons part of the API

TaskRuns and PipelineRuns use the "Reason" field to complement the
value of the "Succeeded" condition. Those values are not part of
the API and are even owned by the underlying resource (pod) in
case of TaskRuns. This makes it difficult to rely on them to
understand that the state of the resource is.

In case of corev1.ConditionTrue, the reason can be used to
distinguish between:
- Successful
- Successful, some parts were skipped (pipelinerun only)

In case of corev1.ConditionFalse, the reason can be used to
distinguish between:
- Failed
- Failed because of timeout
- Failed because of cancelled by the user

In case of corev1.ConditionUnknown, the reason can be used to
distinguish between:
- Just started reconciling
- Validation done, running (or still running)
- Cancellation requested

This is implemented through the following changes:
- Bubble up reasons for taskrun and pipelinerun to the
  v1beta1 API, except for reason that are defined by the
  underlying resource
- Enforce the start reason to be set during condition init

This allows for an additional change in the eventing module: the
condition before and after can be used to decide whether to send
an event at all. If they are different, the after condition now
contains enough information to send the event.

The cloudevent module is extended with ability to send the correct
event based on both status and reason.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
The cloudevent message type strings change from "dev.tekton.event.task.*" to "dev.tekton.event.taskrun.*" for events produced by the `CloudEventPipelineResource`.
```

afrittoli commented:
/kind feature

afrittoli commented:
I cannot reproduce the failure on the timeout test locally - the reason at the end of the pipeline is the wrong one - but it works normally for me.
My suspicion is that status may fail to sync back to etcd occasionally. Perhaps we should do like knative does and retry the update when it fails.

/test pull-tekton-pipeline-integration-tests

jlpettersson commented on pkg/apis/pipeline/v1beta1/pipelinerun_types.go:
```diff
@@ -72,6 +72,21 @@ func (pr *PipelineRun) GetTaskRunRef() corev1.ObjectReference {
+func (pr *PipelineRun) GetStatus() RunsToCompletionStatus {
+	return &pr.Status
+}
+
```
What is the benefit with adding these methods, compared to use the fields directly? e.g. `pr.TypeMeta`, `pr.ObjectMeta` and `pr.Status` ?

Update: Now, I understand - it is part of an interface. Maybe it doesn't have to be part of the interface? I don't know.

afrittoli replied:
Indeed, it's part of the interface. The idea is that a `RunsToCompletion` has a `RunsToCompletionStatus` from which Conditions can be extracted

jlpettersson commented on docs/pipelineruns.md:
```diff
@@ -337,6 +338,70 @@ The `timeout` value is a `duration` conforming to Go's
+False|\[Error message\]|Yes|The `PipelineRun` failed with a permanent error (usually validation).
+False|PipelineRunCancelled|Yes|The `PipelineRun` was cancelled successfully.
+False|PipelineRunTimeout|Yes|The `PipelineRun` timed out.
+False|\[Error message\]|No|The `PipelineRun` encountered an error, but it's still running.
```
Could the two with `[Error message]` but placed next to eachother, it is harder to detect the difference when it is 2 items between them.

afrittoli replied:
Good point, I will fix that

jlpettersson commented on docs/taskruns.md:
```diff
@@ -311,6 +311,22 @@ steps:
+Unknown|TaskRunCancelled|No|The user requested the TaskRun to be cancelled. Cancellation has not be done yet.
+True|Succeeded|Yes|The TaskRun completed successfully.
+False|Failed|Yes|The TaskRun failed because one of the steps failed.
+False|\[Error message\]|Yes|The TaskRun failed with a permanent error (usually validation).
```
Nit: Also here, I find it easier to read if this is moved down to the other `[Error message]`.

afrittoli commented:
/test pull-tekton-pipeline-integration-tests

afrittoli commented:
/test pull-tekton-pipeline-integration-tests

afrittoli commented:
I fixed a small an issue in `InitializeConditions` which was causing CI to fail. I added more unit test coverage for that. The last issue in CI was actually the nop image broken: https://github.com/tektoncd/pipeline/pull/2713

afrittoli commented:
Something is wrong with the cloudevents YAML test
```
ERROR: tests timed out
>> Checking test results for taskrun
ERROR: test poll-for-content-run=SucceededUnknown but should be succeededtrue
ERROR: one or more YAML tests failed
```

jlpettersson commented:
/lgtm

vdemeester commented on pkg/apis/pipeline/v1beta1/run_interface.go:
```diff
@@ -0,0 +1,42 @@
+	IsCancelled() bool
+	HasTimedOut() bool
+	GetRunKey() string
+}
```
@afrittoli any reason for defining those in here and not where it is used ? (it, then, wouldn't need to be exported even :angel:)

afrittoli replied:
Right now it is only used by events, but I was hoping in future to use this to generalise more functions, perhaps even have a base reconciler for `RunToCompletion`.

That would help if we wanted to have dedicated reconcilers for metrics, cloud events or else in the future.

We could have this private to start with and export it in the next PRs if you preferred so.

Since I was adding functions to pipeline run and taskrun types to match this interface, I thought it would make sense to have the interface in the same place, but we could also move it to the controller package.

vdemeester replied:
> We could have this private to start with and export it in the next PRs if you preferred so.

That would be my *preferred* approach indeed :stuck_out_tongue_closed_eyes: I definitely prefer more unexported interfaces (to use the power of the *implicit* part of the interfaces) and small exported interfaces.

afrittoli replied:
done

afrittoli commented:
@n3wscott ping - this may be relevant to you - or you may have feedback on the event type scheme.

pritidesai commented on docs/pipelineruns.md:
```diff
@@ -362,6 +363,70 @@ The `timeout` value is a `duration` conforming to Go's
+to a `TaskRun`, the complete [status of the `TaskrRun`](taskruns.md#monitoring-execution-status) and details
+about `Conditions` that may be associated to a `TaskRun`.
+
+The following example shows an extract from tje `status` field of a `PipelineRun` that has executed successfully:
```
minor nit: typo here tje instead of the

afrittoli replied:
Nice catch, thanks

pritidesai commented on docs/pipelineruns.md:
```diff
@@ -362,6 +363,70 @@ The `timeout` value is a `duration` conforming to Go's
+False|PipelineRunTimeout|Yes|The `PipelineRun` timed out.
+
+When a `PipelineRun` changes status, [events](events.md#pipelineruns) are triggered accordingly.
+
```
thanks @afrittoli for the table, its very helpful and easier to understand, few comments here:

1) it will be easier to understand `PipelineRunCancelled` if both the rows with that reason are in sequence.
2) We are missing `Completed` with status `True` in this table. I think we should include it since this table represents the overall status of `PipelineRun`.
3) May be rename column heading `PipelineRun Status` to description or something similar.
4) This is misleading "The PipelineRun encountered an error, but it's still running." This is happening because [PipelineRun Status](https://github.com/tektoncd/pipeline/blob/be9ac3bf6b27f6beb3f4b7c828ef8388f9b323c3/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L437) is updated with one failure without waiting for all the tasks in `nextRprts` to finish. We can add little more to this description until we fix issue #1680, "The PipelineRun encountered an error, but waiting for already running tasks to finish."

afrittoli replied:
> thanks @afrittoli for the table, its very helpful and easier to understand, few comments here:
> 
>     1. it will be easier to understand `PipelineRunCancelled` if both the rows with that reason are in a sequence.

They are sorted by status. I could sort them by reason, that would make it more readable from a Reason pov, less readable from a status pov.... not sure which is best.
I thought status should go first as it's the primary indicator of the condition.

> 
>     2. We are missing `Completed` with status `True` in this table. I think we should include it since this table represents the overall status of `PipelineRun`.
> 
Nice catch, thank you

>     3. May be rename column heading `PipelineRun Status` to description or something similar.

Fixed
> 
>     4. This is misleading "The PipelineRun encountered an error, but it's still running." This is happening because [PipelineRun Status](https://github.com/tektoncd/pipeline/blob/be9ac3bf6b27f6beb3f4b7c828ef8388f9b323c3/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L437) is updated with one failure without waiting for all the tasks in `nextRprts` to finish. We can add little more to this description until we fix issue #1680, "The PipelineRun encountered an error, but waiting for already running tasks to finish."

That's not the case I was referring to here. 
If any non-permanent error happens during reconcile, the reason is set to the error description, the status is set to failed, but it may be switch back to succeeded if the next reconcile cycle clears the error.

afrittoli replied:
Uhm, I checked this again, and there does not seem to be any transient error reported by the pipelinerun controller, so I will remove this one.
I think I will add the point you made as a general notice to the pipeline run docs.

pritidesai commented on docs/taskruns.md:
```diff
@@ -311,6 +311,23 @@ steps:
+False|TaskRunTimeout|Yes|The TaskRun timed out.
+
+When a `TaskRun` changes status, [events](events.md#taskruns) are triggered accordingly.
+
```
Similar observation as before, may be add a little more description to "The TaskRun encountered an error, but it's still running.", change it "The TaskRun encountered an error, but it's still running since waiting for other tasks or steps (need to check 🔍 ) to finish executing." 🤔

afrittoli replied:
Updated, this is again about transient errors

pritidesai replied:
something like?

```
{"level":"info","logger":"tekton.taskrun-controller.event-broadcaster","caller":"record/event.go:274","msg":"Event(v1.ObjectReference{Kind:\"TaskRun\", Namespace:\"default\", Name:\"pipelinerun-one-failure-two-success-task-c-9fmnd\", UID:\"a788c615-9058-45ae-8dea-de1310343208\", APIVersion:\"tekton.dev/v1beta1\", ResourceVersion:\"4190777\", FieldPath:\"\"}): type: 'Warning' reason: 'Error' Operation cannot be fulfilled on taskruns.tekton.dev \"pipelinerun-one-failure-two-success-task-c-9fmnd\": the object has been modified; please apply your changes to the latest version and try again","commit":"ac439fb","knative.dev/controller":"taskrun-controller"}
```

afrittoli replied:
Yep, failures on update are the only case. Hopefully with the genreconciler we'll see less of that

vdemeester commented on pkg/reconciler/run_interface.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+// RunsToCompletion is implemented by TaskRun and PipelineRun
+type RunsToCompletion interface {
```
Any reason for having it here and not in `events` ? It could then be unexported now ?

afrittoli commented:
/test pull-tekton-pipeline-integration-tests

pritidesai commented on pkg/apis/pipeline/v1beta1/pipelinerun_types.go:
```diff
@@ -200,6 +210,32 @@ type PipelineRunStatus struct {
 }
 
+// PipelineRunSucceededReason represents a reason for the pipeline run "Succeeded" condition
+type PipelineRunSucceededReason string
```
nit: I feel this is coming out too long, can we rename this `PipelineRunSucceededReason` to `PipelineRunReason`, drop `Succeeded`? I understand it represents the `Succeeded` condition which could have status as failure or success depending on the execution but it could be implicit and just call it `PipelineRunReason`. Similar comment to all below reasons 😲  Rename `PipelineRunSucceededReasonStarted` to `PipelineRunReasonStarted` ...

afrittoli replied:
Yeah, I was debating about this one too. 
I thought we might have more conditions in future, and being on the API I wanted to have it unique... but no be fair there would be nothing wrong with having the same reason used in different conditions, so I'll go ahead and shorten the names :)

pritidesai commented on pkg/apis/pipeline/v1beta1/pipelinerun_types.go:
```diff
@@ -200,6 +210,32 @@ type PipelineRunStatus struct {
+	PipelineRunSucceededReasonCancelled PipelineRunSucceededReason = "Cancelled"
+	// PipelineRunSucceededReasonTimedOut is the reason set when the PipelineRun has timed out
+	PipelineRunSucceededReasonTimedOut PipelineRunSucceededReason = "PipelineRunTimeout"
+)
```
I love the way its all in one place now and inside of `types` instead of part of `resources` 🙏

pritidesai commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -102,6 +98,43 @@ type TaskRunStatus struct {
+	TaskRunSucceededReasonCancelled TaskRunSucceededReason = "TaskRunCancelled"
+	// TaskRunSucceededReasonTimedOut is the reason set when the Taskrun has timed out
+	TaskRunSucceededReasonTimedOut TaskRunSucceededReason = "TaskRunTimeout"
+)
```
Similar NIT as above, if we can rename `TaskRunSucceededReason` to `TaskRunReason`, `TaskRunSucceededReasonStarted` to `TaskRunReasonStarted` ...

Nicely collected all reasons under `types` 🙏

pritidesai commented on pkg/apis/pipeline/v1beta1/pipelinerun_types.go:
```diff
@@ -210,13 +246,22 @@ func (pr *PipelineRunStatus) GetCondition(t apis.ConditionType) *apis.Condition
+	// Ensure the started reason is set for the "Succeeded" condition
+	if started {
+		initialCondition := conditionManager.GetCondition(apis.ConditionSucceeded)
+		initialCondition.Reason = TaskRunSucceededReasonStarted.String()
```
Should this be `PipelineRunSucceededReasonStarted` instead of `TaskRunSucceededReasonStarted` ? 🤔  which makes me think, should we combine common PipelineRunReasons and TaskRunReasons e.g. `Started`, `Failed`, `Completed` etc?

afrittoli replied:
Nice catch, thank you!
I think I'd still keep them separated, even if the values are identical.

pritidesai commented on pkg/pod/status.go:
```diff
@@ -114,7 +104,7 @@ func MakeTaskRunStatus(logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev
 		trs.SetCondition(&apis.Condition{
```
Should we create MarkResource* routine for these conditions and use those?

afrittoli replied:
We might do so to reduce boilerplate code, but I would not do it in this PR - https://github.com/tektoncd/pipeline/issues/2770

pritidesai commented on pkg/reconciler/events/cloudevent/cloudevent.go:
```diff
@@ -35,12 +35,32 @@ import (
 const (
+	// TektonTaskRunStartedV1 is sent for TaskRuns with "ConditionSucceeded" "Unknown"
+	// the first time they are picked up by the reconciler
+	TektonTaskRunStartedV1 TektonEventType = "dev.tekton.event.taskrun.started.v1"
```
NIT: rename `TektonTaskRunStarted` to `EventTaskRunStarted` 🤔  and same for all events below 😲

afrittoli replied:
Uhm, maybe `TaskRunStartedEvent`?

afrittoli replied:
Good point, that Tekton in there is redundant

pritidesai commented on pkg/reconciler/events/cloudevent/cloudevent.go:
```diff
@@ -71,27 +118,62 @@ func EventForTaskRun(taskRun *v1beta1.TaskRun) (*cloudevents.Event, error) {
+			eventType = TektonPipelineRunSuccessfulV1
+		}
 	default:
 		return nil, fmt.Errorf("unknown condition for in TaskRun.Status %s", c.Status)
```
NIT: we are now including PipelineRun.Status here as well, may be change this error message to include PipelineRun as well

afrittoli replied:
done

pritidesai commented:
Thanks @afrittoli it looks great, couple of comments above and one more here:

Didnt see Event for PipelineRun Started in (logs)[https://gist.github.com/pritidesai/f90d9057e401bd2b36d84fd1353687eb] after fixing https://github.com/tektoncd/pipeline/pull/2677#discussion_r436158551

afrittoli commented:
> Didnt see Event for PipelineRun Started in [logs](https://gist.github.com/pritidesai/f90d9057e401bd2b36d84fd1353687eb) after fixing [#2677 (comment)](https://github.com/tektoncd/pipeline/pull/2677#discussion_r436158551)

The start event requires https://github.com/tektoncd/pipeline/pull/2545 :)

afrittoli commented:
Rebased and addressed comments @pritidesai @vdemeester

vdemeester commented:
/lgtm cancel
/approve
/meow

afrittoli commented:
yet another rebase :)

pritidesai commented:
thanks @afrittoli, it looks great 🙏 
/lgtm
